File: 2707. Extra Characters in a String/Solution.cpp
#include <string>
#include <unordered_set>
#include <vector>
using namespace std;

class Solution {
public:
    int minExtraChar(string s, vector<string>& dictionary)
    {
        unordered_set<string> dict(dictionary.begin(), dictionary.end());
        vector<int> dp(s.size() + 1, 0);
        for (size_t i = 0; i < s.size(); i++) {
            dp[i + 1] = dp[i] + 1;
            for (int j = i; j >= 0; j--)
                if (dict.contains(s.substr(j, i - j + 1)))
                    dp[i + 1] = min(dp[i + 1], dp[j]);
        }
        return dp.back();
    }
};

File: 2707. Extra Characters in a String/Solution.py
from typing import List


class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        dictionary = set(dictionary)
        dp = [0] * (len(s) + 1)
        for i in range(len(s)):
            dp[i + 1] = dp[i] + 1
            for word in dictionary:
                if s[: i + 1].endswith(word):
                    dp[i + 1] = min(dp[i + 1], dp[i + 1 - len(word)])
        return dp[-1]
